src/Npgsql/NpgsqlActivitySource.cs

```diff
@@ -0,0 +1,69 @@
+﻿using Npgsql.Internal;
+using System;
+using System.Diagnostics;
+using System.Net;
+
+namespace Npgsql
+{
+    static class NpgsqlActivitySource
+    {
+#if NET5_0_OR_GREATER
+        static readonly ActivitySource Source = new("Npgsql");
+
+        public static bool IsEnabled => Source.HasListeners();
+
+        public static IDisposable? CommandStart(NpgsqlConnector connector, string sql)
+        {
+            var settings = connector.Settings;
+            var activity = Source.StartActivity(settings.Database!, ActivityKind.Client);
```

Comment:
```suggestion
            var activity = Source.StartActivity("Execute", ActivityKind.Client);
```

I disagree with the OTel spec here ("When it's otherwise impossible to get any meaningful span name, db.name or the tech-specific database name MAY be used."). For ADO.NET and specifically for `Activity`, I think it would be much more sensible to use `Execute` as the activity name. (A bridge converting `Activity` to OTel could rename the span if it wanted.) But for `ActivitySource`, it seems much better to use a consistent, predictable name for any `ActivityListener` that's interested in database activities (e.g., to potentially filter/distinguish between "Open", "Execute", "Prepare", etc.).

Reply:
Fully agree with you that database name doesn't make a lot of sense for a span name, and making it the same (and more sensible) across multiple providers would have allowed to trace much more rather than just query execution. @roji is there something we can do at the OpenTelemetry level, or is it already agreed on this?

Reply:
Technically, `ActivitySource` is _not_ OpenTelemetry so I think it's OK to define and use our own ADO.NET conventions where they make sense, with the understanding that if `Activity` gets converted to an OTel `Span` that certain changes need to happen. (If we were using `OpenTelemetry.Tracer.Trace` directly, we should follow all OTel conventions. But we're not.)

Reply:
Good question - [here are the OTel guidelines for Span naming](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span). I think we're (rightfully) assuming that we don't do any SQL parsing, and so have no operation or table names.

Here's an extract from the above link, taking an HTTP request as an example:

| Span Name         | Guidance     |
| ----------------- | ------------ |
| `get`             | Too general  |
| `get_account/42`  | Too specific |
| `get_account`     | Good, and account_id=42 would make a nice Span attribute |
| `get_account/{accountId}` | Also good (using the "HTTP route") |

Now, I think the database semantic conventions assume that the only thing you can do with a database is execute SQL on it, which is why the only possible thing (in the absence of table name etc.) is the operation and database names. @bgrainger makes a good point that there are things one can do beyond execute (open, prepare), and these seem to make sense as a low-cardinality verb in the span name (like get for HTTP above).

Another data point is the span name specs in the [HTTP semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md) (also experimental); while servers can have names like `/api/users/{user_id}`, clients know nothing about routing, and so it's recommended to use `HTTP {METHOD_NAME}`. Following this, we could use `DB {verb}` where verb could be Execute/Open/Prepare.

What do you guys think?

(I'd also recommend renaming operation type to mean Execute/Open/Prepare rather than SELECT/INSERT/UPDATE, which seem more like statement types).

(BTW there could be a separate verb for stored procedures, as driver APIs typically make that distinction (e.g. CommandType.StoredProcedure in .NET). In that case we can integrate the sproc name into the span name)

Reply:
I think it's also worth considering how to be a "good"/idiomatic `ActivitySource` in .NET. If you look at https://docs.microsoft.com/en-us/dotnet/core/diagnostics/well-known-event-providers you see that the named providers have a small set of constant CamelCase names. 

Thus, to fit well in the ecosystem. ADO.NET ActivitySource providers should have activity names such as `Execute`, `CallStoredProcedure`, etc. with those Activity names documented in a table somewhere.

It's the responsibility of any system converting `Activity` to an external sink to convert those well-known Activity names into the format the destination system (e.g., OpenTelemetry) expects. (TODO: look up if/how `RequestStart` is converted to `GET Controller/Action/{id}` to make my case.)

Reply:
> Thus, to fit well in the ecosystem. ADO.NET ActivitySource providers should have activity names such as Execute, CallStoredProcedure, etc. with those Activity names documented in a table somewhere.

I think having span names such as Execute or CallStoredProcedure also matches well with the OTel spirit and guidelines - they're not necessarily that different from GET/POST in HTTP (modulu maybe casing).

> It's the responsibility of any system converting Activity to an external sink to convert those well-known Activity names into the format the destination system (e.g., OpenTelemetry) expects. (TODO: look up if/how RequestStart is converted to GET Controller/Action/{id} to make my case.)

So if I'm understanding you right, you're really assuming there would be another component sitting between the ADO.NET provider and the OpenTelemetry listener, performing various conversions between a DiagnosticListener "schema" to a more OTel-compliant one? My expectation was for things to work out-of-the-box for OTel without needing anything else in addition (unless your scenario really has some very specific transformation requirements etc.). It would be good to get some confirmation how things are expected to work.

> TODO: look up if/how RequestStart is converted to GET Controller/Action/{id} to make my case.

Note the difference between HTTP clients - who only know the full URL without any routing info - and HTTP servers which do. Thus the client - and any intermediate component - wouldn't be able to translate `GET Controller/Action/3` into `GET Controller/Action/{id}`, only the server web stack (e.g. ASP.NET) can do that. This is called out in the [HTTP semantic conventions specs](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#name).

More generally, yeah, it does seem like down the road we need to have a better understanding of how ActivitySource and OTel are supposed to interact etc.

Reply:
> So if I'm understanding you right, you're really assuming there would be another component sitting between the ADO.NET provider and the OpenTelemetry listener, performing various conversions between a DiagnosticListener "schema" to a more OTel-compliant one?

I was assuming that, yes (and perhaps it's a bad assumption, or at least one where we're not aligned?). For example, I saw the list of "contrib" projects in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src that mediate between some .NET provider and OpenTelemetry export.

> Note the difference between HTTP clients

Oops, yes, I didn't read carefully to notice that was HttpClient, not ASP.NET Core middleware.

> we need to have a better understanding of how ActivitySource and OTel are supposed to interact etc.

Agreed.

As another example of my thinking, someone might want to submit `Activity` objects to Application Insights (via [TrackDependency](https://docs.microsoft.com/en-us/dotnet/api/microsoft.applicationinsights.telemetryclient.trackdependency?view=azure-dotnet)). Since that's a different "sink", you'd want to follow AppInsights conventions for naming. Again, having a standardised list of ADO.NET Activity names would help that bridge subscribe to the desired events and submit them properly.

For AppInsights, the call might look like this: `TrackDependency("MySQL", $"{conn.DataSource} | {conn.Database}", "SELECT", command.CommandText);` (so the attributes are slightly different than what OTel wants to receive).

We shouldn't unnecessarily use OpenTelemetry names that different bridges might have to "undo" in order to submit this (platform neutral) `ActivitySource` data to other sinks.

But perhaps I'm taking the wrong long-term approach here, and everything _will_ eventually be OpenTelemetry and vendor-neutral, etc. In that world, it would be incumbent on, say, Application Insights to ingest OTel data into _its_ internal format properly, rather than there being a software component that translates from `ActivitySource` to OpenTelemetry or Application Insights (or New Relic or DataDog or anything else).

Reply:
Yeah. My expectation would be that long term things will work out as your describe in the end, i.e. AppInsights would natively consume OpenTelemetry tracing, and that other methods for producing/consuming telemetry would be gradually phased out; after all, the promise of OTel is to make everything "just work" in a standardized way, rather than to require everyone to have to convert their custom telemetry to OTel.

**Some** support for this interpretation may be found in the fact that OTel support ships as part of System.Diagnostics ([see this article](https://devblogs.microsoft.com/dotnet/opentelemetry-net-reaches-v1-0/#library-authors), rather than as some sort of additional plugin (see the comments for some interesting info on AppInsights and other stuff). [This post](https://jimmybogard.com/building-end-to-end-diagnostics-activitysource-and-open/) also details an interest journey from an earlier implementation based on a separate DiagnosticSource Activity package and an OpenTelemetry "translation" functionality to a single, OpenTelemetry-only implementation that just uses ActivitySource as an API for emitting OpenTelemetry.

@shirhatti can you provide some insight here? We're basically trying to figure out we should only keep in mind OpenTelemetry when writing out tracing data, or whether we should design non-OTel tracing data first, which would later be converted to OTel conventions by another component (see the above for more details).